sdk/core/System.ClientModel/src/Pipeline/ClientLoggingPolicy.cs

```diff
@@ -0,0 +1,519 @@
+    public List<string> LoggedHeaderNames { get; internal set; }
+
+    /// <summary>
+    /// TODO.
+    /// </summary>
+    public List<string> LoggedQueryParameters { get; internal set; }
+
+    /// <summary>
+    /// TODO.
+    /// </summary>
+    internal PipelineMessageSanitizer? Sanitizer { get; set; }
+
+    /// <inheritdoc/>
+    public override void Process(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+    {
+        if (!_isLoggingEnabled)
+        {
+            ProcessNext(message, pipeline, currentIndex);
+            return;
+        }
+
+        ProcessSyncOrAsync(message, pipeline, currentIndex, async: false).EnsureCompleted();
+    }
+
+    /// <inheritdoc/>
+    public override async ValueTask ProcessAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+    {
+        if (!_isLoggingEnabled)
+        {
+            await ProcessNextAsync(message, pipeline, currentIndex).ConfigureAwait(false);
+            return;
+        }
+
+        await ProcessSyncOrAsync(message, pipeline, currentIndex, async: true).ConfigureAwait(false);
+    }
+
+    internal virtual void LogRequest()
+    {
+        if (!s_eventSource.IsEnabled())
+        {
+            return;
+        }
+    }
+
+    internal virtual void LogResponse()
+    {
+        if (!s_eventSource.IsEnabled())
+        {
+            return;
+        }
+    }
+
+    private async ValueTask ProcessSyncOrAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex, bool async)
+    {
+        PipelineRequest request = message.Request;
+        string requestId = string.IsNullOrEmpty(_clientRequestIdHeaderName) ? string.Empty : GetRequestIdFromHeaders(request.Headers, _clientRequestIdHeaderName!);
+
+        s_eventSource.Request(request, requestId, _assemblyName, Sanitizer!);
+
+        Encoding? requestTextEncoding = null;
+
+        if (request.Headers.TryGetValue("Content-Type", out string? contentType))
+        {
+            ContentTypeUtilities.TryGetTextEncoding(contentType!, out requestTextEncoding);
+        }
+
+        var logWrapper = new ContentEventSourceWrapper(s_eventSource, _logContent, _maxLength, message.CancellationToken);
+
+        await logWrapper.LogAsync(requestId, request.Content, requestTextEncoding, async).ConfigureAwait(false).EnsureCompleted(async);
+
+        var before = Stopwatch.GetTimestamp();
+
+        try
+        {
+            if (async)
+            {
+                await ProcessNextAsync(message, pipeline, currentIndex).ConfigureAwait(false);
+            }
+            else
+            {
+                ProcessNext(message, pipeline, currentIndex);
+            }
+        }
+        catch (Exception ex)
+        {
+            s_eventSource.ExceptionResponse(requestId, ex.ToString());
+            throw;
+        }
+
+        var after = Stopwatch.GetTimestamp();
+
+        PipelineResponse response = message.Response!;
```

Comment:
Relatively nit-picky comment:

It seems we're interleaving statements related to different functional pieces of the logging routine.

At a high level, it looks like this method does the following:
1. Log request
2. Log request content
3. If exception, log exception and throw
4. Log response (error or non-error)
5. Log response content (wrapped or not)
6. Log delay if elapsed time > threshold

I'm wondering if it would make sense to either refactor the method to look more like just those six functional pieces (a style sometimes called "programming by intention" -- there's a nice discussion of this [here](https://learning.oreilly.com/library/view/essential-skills-for/9780321700469/ch01.html)), or at a minimum, move lines related to wrapping response content closer to where we use the `wrapResponseContent` variable, e.g.  This just makes the code a little easier to read and follow.